src/zino/trapd.py

```diff
@@ -0,0 +1,48 @@
+"""SNMP trap 'daemon' for Zino 2"""
+import asyncio
+import logging
+
+from pysnmp.carrier.asyncio.dgram import udp
+from pysnmp.entity import config, engine
+from pysnmp.entity.rfc3413 import ntfrcv
+
+_logger = logging.getLogger(__name__)
+
+
+class TrapReceiver:
+    """Zino Adapter for SNMP trap reception using PySNMP"""
+
+    def __init__(self, address: str = "0.0.0.0", port: int = 162, loop=None):
+        self.transport = None
+        self.address = address
+        self.port = port
+        self.loop = loop if loop else asyncio.get_event_loop()
+        self.snmp_engine = engine.SnmpEngine()
+        self._communities = set()
+
+    async def open(self):
+        self.transport = udp.UdpTransport(loop=self.loop).openServerMode((self.address, self.port))
+        # This attribute needs to be awaited to ensure the socket is really opened,
+        # unsure of why PySNMP doesn't do this, or how else it's supposed to be achieved
+        await self.transport._lport
+        _logger.info("Listening for incoming SNMP traps on %r", (self.address, self.port))
+        config.addTransport(self.snmp_engine, udp.domainName + (1,), self.transport)
+
+        ntfrcv.NotificationReceiver(self.snmp_engine, self.trap_received)
+        self.snmp_engine.transportDispatcher.jobStarted(1)  # this job would never finish
+
+    def add_community(self, community: str):
```

Comment:
Will Zino 2 support SNMP 3? Might be easier learning how to doit *here* then use that kmowledge in Nav than the other way around.

Reply:
We have no (current) plans to support SNMPv3, just to have the same feature set as Zino 1.

Multiple "systems" can be configured on the SNMP engine.  This code just adds a v1 system, which is the basic "community" for authentication, which works for both v1 and v2.  PySNMP also has the option to add v3 systems to the engine, which are more involved.

However, none of this is very relevant to NAV, since NAV uses an entirely different SNMP library.